### ISSUE TITLE
Proposal: Make WellKnownMutability customizable via compiler args

### ISSUE THREAD

**Issue by hzsweers** [google/error-prone#512]
We have internal clones of guava's immutable collections (android, we don't want the full library packaged). The result though is that some of our autovalue classes using them can't be denoted as `@Immutable` because it only looks for guava's. It would be ideal if this could be more customizable. Thoughts?

**Comment by cushon** [comment]
Can you add `@Immutable` to the declarations of those immutable collection types? The whitelist was intended for cases where we can't add the annotations, like the JDK libraries.

**Comment by hzsweers** [comment]
I suppose we could, but it was just one example. Another example might be a class with lazy inits that are final once initialized, which we wouldn't be able to put the annotation on either.

**Comment by cushon** [comment]
The analysis trusts fields annotated with [`@LazyInit`](http://errorprone.info/api/latest/com/google/errorprone/annotations/concurrent/LazyInit.html).

More generally, if there are patterns that you know are safe but that the analysis doesn't handle, there's always `@SuppressWarnings("Immutable")`.

**Comment by hzsweers** [comment]
Huh, I didn't know about that. Wonder if I should propose generating those on memoized autovalue fields. Alright, thanks for the insight!

**Comment by JakeWharton** [comment]
I don't think this is resolved. The inability to alter this list is a source of constant pain for us as well. We pull in libraries with immutable types and also consume generated code that creates immutable types that we don't have the ability to annotate. I'm not going to make pull requests adding a bunch of random library types to the list but I also don't want to lose the change just do to the lack of an extensibility point.

**Comment by hzsweers** [comment]
We're seeing more cases now, I think it's worth a discussion at least

**Comment by hzsweers** [comment]
Another example: interfaces from external libraries aren't annotated as Immutable, but are also not mutable interfaces. This is a pretty common occurrence

**Comment by eaftan** [comment]
@epmjohnston is actively working on this in #524.

**Comment by lowasser** [comment]
(also, on a separate note, Guava is working on making the Immutable collections much better for Android for future releases)

**Comment by epmjohnston** [comment]
Immutability checkers are now customizable using the brand new Error Prone command line args! (see docs at http://errorprone.info/docs/flags)

Provide lists of your own known immutable and known unsafe classes, like so:
- "-XepOpt:Immutable:KnownImmutable=some.immutable.Clazz,other.safe.Object"
- "-XepOpt:Immutable:KnownUnsafe=some.mutable.Clazz,other.unsafe.Object"

Try it out, and be sure to leave feedback to let me know how they work out for you!

**Comment by JakeWharton** [comment]
Does this require specifying every class, or can I supply a base class and
have all subclasses be marked? I'm specifically dealing with a
code-generation library so there's a few thousand individual classes here.

**Comment by epmjohnston** [comment]
This requires specifying every class individually. We hesitate to mark subclasses. Though with good reason perhaps we could add another flag with this behavior, or some other solution.

What exactly is the use case? Do you want to mark the generation library or the generated code as immutable? If the latter, can you change the code generator to add `@Immutable` to the generated classes?

**Comment by JakeWharton** [comment]
It's the generated code that we want to mark. I'm hesitant to mark the generated code directly since it's an open source project and so there's a bunch of consumers that we don't know about with regard to the added dependency they we'd need to compile. That said, I also understand why you're hesitant about the ability to mark generated code since it opens up the potential for subclasses to willfully violate that immutability.

Just to be clear, only your `@Immutable` annotation works? Or any named `@Immutable`?

**Comment by epmjohnston** [comment]
Currently, only our `@Immutable` annotation is checked.

One solution we're considering: add a boolean flag to enable checking any annotation named `Immutable`. Then you could annotate the generated classes with `javax.concurrent.Immutable`, set the flag, and have Error Prone check those classes for immutability. Would that suit your use case?

**Comment by epmjohnston** [comment]
ping @JakeWharton in case you didn't see this

**Comment by JakeWharton** [comment]
Sorry I'm changing jobs so things are crazy. I'll look soon!

**Comment by pyricau** [comment]
I think supporting custom `@Immutable` would be nice, very similar to how libraries deal with `@Nullable`. I wouldn't have a boolean (or I'd have that behavior on by default). The likeliness of someone using a rogue `@Immutable` that doesn't mean what we think it means seems pretty low.
